diff --git a/erts/emulator/beam/beam_load.c b/erts/emulator/beam/beam_load.c
index 00179cd..8fc188d 100644
--- a/erts/emulator/beam/beam_load.c
+++ b/erts/emulator/beam/beam_load.c
@@ -44,6 +44,9 @@
 #include "erl_process_dict.h"
 #include "erl_unicode.h"
 #include "beam_file.h"
+#ifdef __rtems__
+#include "erl_gc.h"
+#endif

 Uint erts_total_code_size;

@@ -133,6 +136,11 @@ erts_prepare_loading(Binary* magic, Process *c_p, Eterm group_leader,
     stp->module = *modp;
     stp->group_leader = group_leader;

+#ifdef __rtems__
+    erts_fprintf(stderr, "Loading module: %T\n", stp->module);
+    /*erts_garbage_collect(c_p, 0, NULL, 0);*/
+#endif
+
 #if defined(LOAD_MEMORY_HARD_DEBUG) && defined(DEBUG)
     erts_fprintf(stderr,"Loading a module\n");
 #endif
