diff --git a/erts/emulator/beam/beam_load.c b/erts/emulator/beam/beam_load.c
index d051085..3a64573 100644
--- a/erts/emulator/beam/beam_load.c
+++ b/erts/emulator/beam/beam_load.c
@@ -131,6 +131,8 @@ erts_prepare_loading(Binary* magic, Process *c_p, Eterm group_leader,
     stp->module = *modp;
     stp->group_leader = group_leader;

+    erts_fprintf(stderr, "Loading module: %T\n", stp->module);
+
 #if defined(LOAD_MEMORY_HARD_DEBUG) && defined(DEBUG)
     erts_fprintf(stderr,"Loading a module\n");
 #endif
